### ISSUE TITLE
Demo for ASR using HDF Dataset.

### ISSUE THREAD

**Issue by yanghongjiazheng** [rwth-i6/returnn#322]
Dear authors, thanks for your great work. We want to train a ASR model over huge corpus. To accelerate this, we've tried multi-gpu training. However, the result turn me down. I also find that you've offered HDF dataset for training-accelerating. Can you provide a simple demo?  Thanks a lot!!

**Comment by albertz** [comment]
Note that StackOverflow is a better place to ask for help (to ask questions in general). The GitHub issue tracker is about issues in RETURNN. I think there is not really any bug or problem here, right?

But anyway:

How did you use multi-GPU?
I currently would recommend a setting like this:
```
horovod_dataset_distribution = "random_seed_offset"
horovod_reduce_type = "param"
horovod_param_sync_time_diff = 100.
```
Please read the documentation and make sure you understand what this means.

For dumping your dataset to HDF, you can use the `hdf_dump.py` tool. But please first read the documentation to understand whether the dataset is really the bottleneck for your case. Is it?

**Comment by albertz** [comment]
@JackTemaki might want to extend the documentation.

**Comment by yanghongjiazheng** [comment]
Yes, sure. I have go over all issues relating to multi-gpu training. And my setting is exactly what you've given above.  The steps for single-gpu and 4-gpus is  right. But the computing time for 4-gpu is only 2%~4%. 
`pretrain epoch 4, finished after 0 steps, 0:00:00 elapsed (0.0% computing time)
pretrain epoch 4 'dev' eval, finished after 3 steps, 0:04:59 elapsed (2.9% computing time)
pretrain epoch 5, finished after 0 steps, 0:00:15 elapsed (0.0% computing time)
pretrain epoch 5 'dev' eval, finished after 3 steps, 0:04:57 elapsed (2.8% computing time)
pretrain epoch 6, finished after 3 steps, 0:04:16 elapsed (5.0% computing time)
pretrain epoch 6 'dev' eval, finished after 3 steps, 0:04:57 elapsed (2.8% computing time)
pretrain epoch 7, finished after 3 steps, 0:05:12 elapsed (3.0% computing time)
pretrain epoch 7 'dev' eval, finished after 3 steps, 0:04:59 elapsed (2.8% computing time)
pretrain epoch 8, finished after 2 steps, 0:04:11 elapsed (2.9% computing time)`
I'll try hdf_dump.py. But I think the transformation from origin files  to hdf files consuming time itself. Can you give a example about how much time it will cost when transferring 60G wav files to hdf files?  And the batch size for "GTX 1080 with 11 GB RAM" is 20000, so different machines have  different best set for batch size?  All should I do is to find a threshold, that makes the computing time up to 90%?   Then , when training with multi-gpu, should I keep the batch size same as that for single gpu or to turn lower or upper? Thanks a lot .

**Comment by albertz** [comment]
Your epochs look very strange. There are 0 steps for epoch 4 and 5, and then only 3 steps for the next epochs. You normally would expect something like 1000-100k steps per epoch. I assume sth is wrongly configured there.
* Maybe the dataset itself (check with the `dump-dataset.py` tool). An epoch usually should consist of a couple of 1000 sequences.
* Or your mini-batch settings. What is your batch size? Num seqs? Etc. Follow some of our published configs.
  To better debug this, it will be helpful to increase some further log messages. Set this:
  ```
  log_verbosity = 5
  tf_log_memory_usage = True
  log_batch_size = True
  ```
  See [here](https://github.com/rwth-i6/returnn-experiments/blob/master/2020-rnn-transducer/configs/base2.conv2l.specaug4a.ctc.devtrain.config) for a full example.

> But I think the transformation from origin files to hdf files consuming time itself.

Yes, sure. But this would be a one-time thing. But anyway, with `horovod_dataset_distribution = "random_seed_offset"` (which I would recommend to use) you should not need this anyway.

> Can you give a example about how much time it will cost when transferring 60G wav files to hdf files?

This is hard to estimate, this really depends on your hardware. I assume a couple of hours.

But I anyway would recommend to convert it to Ogg instead, and then use the `OggZipDataset`. That should be much faster already.

What dataset do you use currently? And what type of features?

> different machines have different best set for batch size?

Yes. But having a bigger batch size is not always faster or better. You could end up with more zero-padding in a batch (depends how you shuffle & sort it; I guess you use `laplace`?). Or the convergence rate might also be slower.

> Then , when training with multi-gpu, should I keep the batch size same as that for single gpu or to turn lower or upper?

Should be the same as before. But the convergence rate can be different. You might need to play around with it.

**Comment by yanghongjiazheng** [comment]
OK , I got this. The epochs showed above is pretraining stage. And I only use the libri-dataset train-clean 100 for a test. So the step is not normal.
Besides, the batch size is 5000，num_sqs is 200
Our gpu is P-40

--------------原始邮件--------------
发件人："Albert Zeyer "<notifications@github.com&gt;;
发送时间：2020年7月14日(星期二) 晚上7:32
收件人："rwth-i6/returnn" <returnn@noreply.github.com&gt;;
抄送："yanghongjiazheng "<yangjia@bupt.edu.cn&gt;;"Author "<author@noreply.github.com&gt;;
主题：Re: [rwth-i6/returnn] Demo for ASR using HDF Dataset. (#322)
-----------------------------------

 

 
Your epochs look very strange. There are 0 steps for epoch 4 and 5, and then only 3 steps for the next epochs. You normally would expect something like 1000-100k steps per epoch. I assume sth is wrongly configured there.
  
Maybe the dataset itself (check with the dump-dataset.py tool). An epoch usually should consist of a couple of 1000 sequences.
 
Or your mini-batch settings. What is your batch size? Num seqs? Etc. Follow some of our published configs.
 To better debug this, it will be helpful to increase some further log messages. Set this: log_verbosity = 5 tf_log_memory_usage = True log_batch_size = True  See here for a full example.
   
But I think the transformation from origin files to hdf files consuming time itself.
  
Yes, sure. But this would be a one-time thing. But anyway, with horovod_dataset_distribution = "random_seed_offset" (which I would recommend to use) you should not need this anyway.
  
Can you give a example about how much time it will cost when transferring 60G wav files to hdf files?
  
This is hard to estimate, this really depends on your hardware. I assume a couple of hours.
 
But I anyway would recommend to convert it to Ogg instead, and then use the OggZipDataset. That should be much faster already.
 
What dataset do you use currently? And what type of features?
  
different machines have different best set for batch size?
  
Yes. But having a bigger batch size is not always faster or better. You could end up with more zero-padding in a batch (depends how you shuffle &amp; sort it; I guess you use laplace?). Or the convergence rate might also be slower.
  
Then , when training with multi-gpu, should I keep the batch size same as that for single gpu or to turn lower or upper?
  
Should be the same as before. But the convergence rate can be different. You might need to play around with it.

**Comment by albertz** [comment]
> OK , I got this. The epochs showed above is pretraining stage. And I only use the libri-dataset train-clean 100 for a test.

Ok. I assume you also use the `epoch_wise_filter`? Note that if you have 0 steps in one epoch, then there was not a single sequence and it did nothing.

Which config did you follow? A recent one for LibriSpeech is e.g. [here](https://github.com/rwth-i6/returnn-experiments/blob/master/2019-asr-e2e-trafo-vs-lstm/librispeech/base2.conv2l.specaug.curric3.config).

Do you have the very latest RETURNN version? `horovod_dataset_distribution` and also `horovod_param_sync_time_diff` only work with the latest version.

I also wonder about your high time for only 3 steps (~5 minutes). This looks like sth else is wrong as well.

Can you maybe share your config?

**Comment by yanghongjiazheng** [comment]
The config file is attached.&nbsp; And the returnn is not the latest, but the&nbsp;&nbsp;horovod_dataset_distribution&nbsp; did work.&nbsp;


commit c28b6f4f3431495f0dfa0bc04d8753cba4a81950 (HEAD -&gt; master, origin/master, origin/HEAD)
Author: Nick Rossenbach <rossenbach@i6.informatik.rwth-aachen.de&gt;
Date:&nbsp; &nbsp;Wed May 29 03:17:41 2019 +0200



And the step showed before is one log file for the 4-gpu training, and I didn't use the hdfdataset





&nbsp;
------------------&nbsp;Original&nbsp;------------------

**Comment by albertz** [comment]
> The config file is attached.

I don't see it. You need to attach it to GitHub. Or maybe put it into a Gist and link it here.
(Btw, also better answer here directly. Your mail client seems to mess up the answer.)

> And the returnn is not the latest, but the `horovod_dataset_distribution` did work

The commit you referenced (c28b6f4, from May 2019), this is your RETURNN version?
How do you know `horovod_dataset_distribution` works? This definitely cannot work. It was just implemented this month. As well as for the other Horovod option.

Please check with the latest RETURNN version.

**Comment by yanghongjiazheng** [comment]
Ok， thanks a lot!

发件人："Albert Zeyer "<notifications@github.com&gt;;
发送时间：2020年7月14日(星期二) 晚上9:21
收件人："rwth-i6/returnn" <returnn@noreply.github.com&gt;;
抄送："yanghongjiazheng "<yangjia@bupt.edu.cn&gt;;"Author "<author@noreply.github.com&gt;;
主题：Re: [rwth-i6/returnn] Demo for ASR using HDF Dataset. (#322)
-----------------------------------

 

  
The config file is attached.
  
I don't see it. You need to attach it to GitHub. Or maybe put it into a Gist and link it here.
 (Btw, also better answer here directly. Your mail client seems to mess up the answer.)
  
And the returnn is not the latest, but the horovod_dataset_distribution did work
  
The commit you referenced (c28b6f4, from May 2019), this is your RETURNN version?
 How do you know horovod_dataset_distribution works? This definitely cannot work. It was just implemented this month. As well as for the other Horovod option.
 
Please check with the latest RETURNN version.

**Comment by yanghongjiazheng** [comment]
Dear author. I have updated the latest version Returnn. And set the parameters relating to the multi-gpu on. However, the Librispeech  testing result shows that multi-gpu training doesn't help improve the training efficiency .
`        one gpu train:       finished after 888 steps,0:22:05 elapsed (79.3% computing time)

          two gpus train:    finished after 888 steps, 0:41:54 elapsed (50.4% computing time)    

                                       finished after 888 steps, 0:41:54 elapsed (50.4% computing time)

          three gpus train:  finished after 888 steps, 1:05:52 elapsed (33.8% computing time)

                                        finished after 888 steps,  1:05:52 elapsed (33.5% computing time)

                                         finished after 888 steps, 1:05:52 elapsed (33.8% computing time) `

**Comment by albertz** [comment]
What settings do you use exactly now? Can you post your config and the log in a Gist?

**Comment by yanghongjiazheng** [comment]
And also to do a simple test. I've transformed the librispeech train-clean-100 to a hdf dataset. But I have no idea about how to use it. Can you give me some help?

**Comment by albertz** [comment]
> I've transformed the librispeech train-clean-100 to a hdf dataset. But I have no idea about how to use it. Can you give me some help?

Where exactly is the problem? Or what exactly do you not know? How to use a HDF file? You just use the `HDFDataset` for that. In your config sth like this:
```
train = {"class": "HDFDataset", "files": ["file.hdf"]}
```

However, with `horovod_dataset_distribution = "random_seed_offset"`, it shouldn't really be so relevant.

But as I said, can you post your log + config to a Gist? That would help in better understanding what goes wrong there.

**Comment by yanghongjiazheng** [comment]
https://gist.github.com/yanghongjiazheng/72382be106403df79615591934898310
Here is my config and one training log of the 3-gpus training process

**Comment by yanghongjiazheng** [comment]
To do a simple test. I only use librispeech train-clean-100 as training data and we keep the dev and test as usual.

**Comment by albertz** [comment]
I see that you use `horovod_param_sync_step = 100`. Did you also tried with `horovod_param_sync_time_diff = 100.` as explained above?

Where do you have your data files? I.e. where is `data/dataset`? This might be your bottleneck. I.e. just the loading of the files from there. Now with N parallel processes, it means that N processes read from there at the same time, so you would need N times more bandwidth (if you want to have perfect speedup).

I would recommend to copy the files to a local disk. That will be much faster than any network filesystem.
And also, convert it to Ogg (if you do not have already). Then the files become much smaller, and that also will speedup the file reading a lot.

**Comment by yanghongjiazheng** [comment]
Thanks a lot.  The data files is right on a local disk. And the path is `returnn-train/data/dataset`.  And the returnn path is `returnn-train/returnn`. So I think  maybe the bottleneck isn't loading of the files from there. But I didn't convert it to Ogg.

**Comment by yanghongjiazheng** [comment]
I‘m wondering if you can provide your time saving performance using multi-gpus?

**Comment by yanghongjiazheng** [comment]
I've just tried the horovod_param_sync_time_diff = 100 . And the performance is just same as before

**Comment by albertz** [comment]
I get a perfect speed up with these settings. Computing time, and absolute time almost do not change at all for me.
Computing time should always be close to 100% (for train at least) (and that's what I usually have). If you do not have that, it means that there is some bottleneck. Just check the code for that, it should be clear how this is computed. So in your case, it seems huge amounts of time are spent somewhere else than in `sess.run`. The question is where exactly. Most likely, this is the dataset. You could use some Python profiler to check (for example [Py-Spy](https://github.com/benfred/py-spy)). Or add some further timings in the code.
So, if I'm right, and your dataset is slow, the question is why. Maybe the file loading is not slow (because it is already on local disk). But then maybe the feature extraction (MFCC) is slow. This is done via librosa, i.e. via Numpy, on CPU. If your CPU is busy by other processes, this might be a bottleneck. If that is the case, the HDF should definitely speed it up, as this would not do feature extraction anymore.

**Comment by yanghongjiazheng** [comment]
My training corpus are librispeech datasets. And all the audios are flac format.  Are you the same ?

**Comment by albertz** [comment]
I use Ogg files for LibriSpeech. But in any case, I did not use multi-GPU training that much for Librispeech yet. I have another dataset for my current multi-GPU experiments (`ExternSprintDataset`, which is similar to `HDFDataset`), which does not do the feature extraction on-the-fly, i.e. it just loads the features (MFCC or GT) from disk. So there would not really be a CPU bottleneck here, because this mostly does nothing, except of copying data around. In your case, you do feature extraction on-the-fly. So if that is the bottleneck, then the `HDFDataset` should improve things a lot for you. Did you try that?

**Comment by yanghongjiazheng** [comment]
Yes  I do feature extraction on the fly. I did try "hdf_dump.py" to convert Librispeech to HDFDataset. But I didn't use it in right way . I just run `python3 returnn/toos/hdf_dump.py    returnn.config    train-clean-200.hdf` . But it didn't  work.  Can you offfer  some help or config file to do this.

**Comment by yanghongjiazheng** [comment]
What does the parameter  'epoch' in the hdf_dump.py  use for ?    Does this script convert all the data used in the 0~$epoch epochs to hdfdataset?

**Comment by albertz** [comment]
> I just run `python3 returnn/toos/hdf_dump.py returnn.config train-clean-200.hdf` . But it didn't work.

What do you mean, it didn't work? What happened?

You might want to edit the config for `hdf_dump.py`. Or better create a separate config, just for that. Because e.g. you would want to remove `epoch_wise_filter`, `seq_ordering`, `partition_epoch`.

The script just dumps the data from one epoch (the first epoch) to a HDF file.

**Comment by yanghongjiazheng** [comment]
Thanks a lot.  I tried `python3 returnn/toos/hdf_dump.py --epoch 200 returnn.config train-clean-200.hdf' And it did work.  `    two gpus:   
 train epoch 230, finished after 954 steps, 0:20:23 elapsed (99.3% computing time)
train epoch 230, finished after 954 steps, 0:20:23 elapsed (99.1% computing time)
    one gpus:    
train epoch 230, finished after 954 steps, 0:19:23 elapsed (99.1% computing time)`

And I still have two questions , hoping you can give some advice. First of all, the hdfdata set holds the origin audios or mfcc features?   And if it hold the audios, can I set this config for training data`train={"class": "HDFDataset", "files": ["test_epoch_1.hdf"],"audio":{"random_permute":True}}` to do some data augmentation?

**Comment by yanghongjiazheng** [comment]
I find that a thread named `self.data_provider.start_threads(session=sess)`  starts in the TFEngine.py.  Are the numbers of threads  Configurable？  I think maybe this will help improve the performance too.

**Comment by albertz** [comment]
> Thanks a lot. I tried `python3 returnn/toos/hdf_dump.py --epoch 200 returnn.config train-clean-200.hdf` And it did work.

Why did you use `--epoch 200`? That does not really make sense.
And you edited the `returnn.config` and removed `epoch_wise_filter`, `seq_ordering`, `partition_epoch`, right?

> train epoch 230, finished after 954 steps, 0:20:23 elapsed (99.1% computing time)

Yes, this looks much better.
So this means that your CPU was too slow in the other case.

> First of all, the hdfdata set holds the origin audios or mfcc features?

`hdf_dump.py` just dumps whatever comes out of the dataset which you provide in the config. So this dumps MFCC.
This is what I explained. Otherwise the whole thing would not make sense. This is the actual point why you do this at all. Otherwise you again would have the same problem as before.

> And if it hold the audios, can I set this config for training data`train={"class": "HDFDataset", "files": ["test_epoch_1.hdf"],"audio":{"random_permute":True}}` to do some data augmentation?

No.

> I find that a thread named `self.data_provider.start_threads(session=sess)` starts in the TFEngine.py. Are the numbers of threads Configurable？ I think maybe this will help improve the performance too.

This is only a single thread, and this single thread is necessary. There is nothing to configure about that.
